File: app/src/main/java/com/themovielist/moviedetail/MovieDetailPresenter.java
package com.themovielist.moviedetail;


import android.support.annotation.NonNull;

import com.themovielist.base.BasePresenterImpl;
import com.themovielist.model.MovieModel;
import com.themovielist.model.MovieReviewModel;
import com.themovielist.model.MovieTrailerModel;
import com.themovielist.repository.ArrayRequestAPI;
import com.themovielist.repository.movie.MovieRepository;

import java.util.List;

import javax.inject.Inject;

import timber.log.Timber;

public class MovieDetailPresenter extends BasePresenterImpl implements MovieDetailContract.Presenter {
    private MovieDetailContract.View mView;
    private ArrayRequestAPI<MovieReviewModel> mMovieReviewRequest;
    private List<MovieTrailerModel> mMovieTrailerList;
    private int mMovieId;

    @Inject
    MovieDetailPresenter(@NonNull MovieRepository movieRepository) {
        super(movieRepository);
    }

    @Override
    public void start(MovieModel movieModel) {
        mView.showMovieDetail(movieModel);

        // If it is in the database, means that is favorite.
        getMMovieRepository().getMovieDetailById(movieModel.getId()).subscribe(movieModel1 -> mView.setFavoriteButtonState(true));

        loadMovieReviews(movieModel.getId());

        loadMovieTrailers(movieModel.getId());

        mMovieId = movieModel.getId();
    }

    private void loadMovieTrailers(int movieId) {
        mView.showLoadingTrailersIndicator();
        getMMovieRepository().getTrailersByMovieId(movieId).subscribe(
                this::handleMovieTrailerRequestSuccess,
                throwable -> {
                    Timber.e(throwable, "An error occurred while tried to get the movie trailers");
                    mView.showErrorMessageLoadTrailers();
                });
    }

    private void loadMovieReviews(int movieId) {
        mView.showLoadingReviewsIndicator();
        getMMovieRepository().getReviewsByMovieId(1, movieId).subscribe(
                this::handleMovieReviewRequestSuccess,
                throwable -> {
                    Timber.e(throwable, "An error occurred while tried to get the movie reviews");
                    mView.showErrorMessageLoadReviews();
                });
    }

    private void handleMovieTrailerRequestSuccess(List<MovieTrailerModel> movieTrailerModels) {
        if (movieTrailerModels.size() == 0) {
            mView.showEmptyTrailerListMessage();
            mView.setShowAllTrailersButtonVisibility(false);
            return;
        }

        mMovieTrailerList = movieTrailerModels;
        if (mMovieTrailerList.size() > 2) {
            mView.showMovieTrailer(mMovieTrailerList.subList(0, 2));
            mView.setShowAllTrailersButtonVisibility(true);
        } else {
            mView.showMovieTrailer(mMovieTrailerList);
            mView.setShowAllTrailersButtonVisibility(false);
        }
    }

    private void handleMovieReviewRequestSuccess(ArrayRequestAPI<MovieReviewModel> movieReviewModelArrayRequestAPI) {
        if (movieReviewModelArrayRequestAPI.getResults().size() == 0) {
            mView.showEmptyReviewListMessage();
            mView.setShowAllReviewsButtonVisibility(false);
            return;
        }
        mMovieReviewRequest = movieReviewModelArrayRequestAPI;
        if (movieReviewModelArrayRequestAPI.getResults().size() > 2) {
            mView.showMovieReview(mMovieReviewRequest.getResults().subList(0, 2));
            mView.setShowAllReviewsButtonVisibility(true);
        } else {
            mView.showMovieReview(mMovieReviewRequest.getResults());
            mView.setShowAllReviewsButtonVisibility(false);
        }
    }

    @Override
    public void setView(MovieDetailContract.View view) {
        mView = view;
    }

    @Override
    public void removeFavoriteMovie(MovieModel movieModel) {
        getMMovieRepository().removeFavoriteMovie(movieModel).subscribe(
                mView::showSuccessMessageRemoveFavoriteMovie,
                throwable -> {
                    Timber.e(throwable, "An error occurred while tried to remove the favorite movie");
                    mView.showErrorMessageRemoveFavoriteMovie();
                    mView.setFavoriteButtonState(true);
                });
    }

    @Override
    public void saveFavoriteMovie(MovieModel movieModel) {
        getMMovieRepository().saveFavoriteMovie(movieModel).subscribe(
                mView::showSuccessMessageAddFavoriteMovie,
                throwable -> {
                    Timber.e(throwable, "An error occurred while tried to add the favorite movie");
                    mView.showErrorMessageAddFavoriteMovie();
                    mView.setFavoriteButtonState(false);
                });
    }

    @Override
    public void showAllReviews() {
        mView.showAllReviews(mMovieReviewRequest.getResults(), mMovieReviewRequest.hasMorePages());
    }

    @Override
    public void showAllTrailers() {
        mView.showAllTrailers(mMovieTrailerList);
    }

    @Override
    public void tryToLoadTrailersAgain() {
        loadMovieTrailers(mMovieId);
    }

    @Override
    public void tryToLoadReviewAgain() {
        loadMovieReviews(mMovieId);
    }
}

File: app/src/main/java/com/themovielist/moviedetail/MovieDetailPresenter.kt
package com.themovielist.moviedetail


import com.themovielist.base.BasePresenterImpl
import com.themovielist.model.MovieModel
import com.themovielist.model.MovieWithGenreModel
import com.themovielist.model.response.MovieDetailResponseModel
import com.themovielist.repository.movie.CommonRepository
import com.themovielist.repository.movie.MovieRepository
import timber.log.Timber
import javax.inject.Inject

class MovieDetailPresenter @Inject
internal constructor(movieRepository: MovieRepository, private val mCommonRepository: CommonRepository) : BasePresenterImpl(movieRepository), MovieDetailContract.Presenter {
    private lateinit var mView: MovieDetailContract.View
    private lateinit var mMovieWithGenreModel: MovieWithGenreModel
    private lateinit var mMovieDetailResponseModel: MovieDetailResponseModel
    private var isFavorite = false

    override fun start(movieModel: MovieModel) {
        mView.showLoadingMovieDetailIndicator()
        mCommonRepository.getAllGenres().subscribe({ genreMap ->
            val genreList = mCommonRepository.fillMovieGenresList(movieModel, genreMap)
            mMovieWithGenreModel = MovieWithGenreModel(movieModel, genreList)
            mView.showMovieInfo(mMovieWithGenreModel)

            fetchMovieDetailInfo()
        }, { error ->
            mView.showErrorLoadingMovieDetail(error)
        })

        mMovieRepository.isMovieFavorite(movieModel.id).subscribe({
            mView.setFavoriteButtonState(it)
            isFavorite = it
            Timber.i("IS MOVIE FAVORITE: $it")
        })
    }

    private fun fetchMovieDetailInfo() {
        mView.showLoadingMovieDetailIndicator()
        mMovieRepository.getMovieDetail(mMovieWithGenreModel.id)
                .subscribe(this::handleGetMovieDetailResponseSuccess, { error ->
                    mView.showErrorLoadingMovieDetail(error)
                })
    }

    private fun handleGetMovieDetailResponseSuccess(movieDetailResponseModel: MovieDetailResponseModel) {
        mMovieDetailResponseModel = movieDetailResponseModel
        mView.showMovieDetailInfo()

        mView.bindMovieReviewInfo(mMovieDetailResponseModel.reviewsResponseModel.results)
        mView.bindMovieTrailerInfo(mMovieDetailResponseModel.trailerResponseModel.trailerList)

        val hourMinute = convertRuntimeToHourMinute(movieDetailResponseModel.runtime)
        mView.showMovieRuntime(hourMinute)

        mView.hideLoadingMovieDetailIndicator()
    }

    private fun convertRuntimeToHourMinute(runtime: Int): Pair<Int, Int> {
        val hours = runtime / 60
        val minutes = runtime % 60
        return Pair(hours, minutes)
    }

    override fun setView(view: MovieDetailContract.View) {
        mView = view
    }

    override fun showAllReviews() {
        mView.showAllReviews(mMovieDetailResponseModel.reviewsResponseModel.results, mMovieDetailResponseModel.reviewsResponseModel.hasMorePages())
    }

    override fun showAllTrailers() {
        mView.showAllTrailers(mMovieDetailResponseModel.trailerResponseModel.trailerList)
    }

    fun toggleFavoriteMovie() {
        val request = if (isFavorite) {
            mMovieRepository.removeFavoriteMovie(mMovieWithGenreModel)
        } else {
            mMovieRepository.saveFavoriteMovie(mMovieWithGenreModel)
        }

        mView.setFavoriteButtonEnabled(false)
        request
                .doOnTerminate {
                    mView.setFavoriteButtonState(isFavorite)
                    mView.setFavoriteButtonEnabled(true)
                }
                .subscribe(
                        {
                            if (isFavorite) {
                                mView.showSuccessMessageRemoveFavoriteMovie()
                            } else {
                                mView.showSuccessMessageAddFavoriteMovie()
                            }
                            isFavorite = !isFavorite
                            mView.setFavoriteButtonState(isFavorite)
                            mView.dispatchFavoriteMovieEvent(mMovieWithGenreModel, isFavorite)
                        }
                ) { throwable ->
                    Timber.e(throwable, "An error occurred while tried to remove the favorite movie")
                    mView.showErrorMessageRemoveFavoriteMovie()
                }
    }

    override fun tryFecthMovieDetailAgain() = fetchMovieDetailInfo()
}
